nav2_costmap_2d/plugins/voxel_layer.cpp

```diff
@@ -169,15 +169,15 @@ void VoxelLayer::reset()
 {
   voxel_pub_->on_deactivate();
   deactivate();
-  resetMaps();
+  clearMap();
   activate();
   undeclareAllParameters();
   voxel_pub_->on_activate();
 }
 
-void VoxelLayer::resetMaps()
+void VoxelLayer::clearMap()
 {
-  Costmap2D::resetMaps();
```

Comment:
As far as I can tell from this diff, you didnt remove resetMaps from costmap2D. Should you? What’s the difference between clear and reset?

Reply:
You also called the costmap2D directly not the obstacle layer this derives from that also needs to do useful “stuff”

Reply:
reset - calls resetMaps, initializes everything including parameters, subscribers and publishers
clearMap - calls resetMaps or/and sets the necessary flags

Reply:
I will test this to simplify clearMap in voxelLayer.

VoxelLayer::clearMap()
{
  resetMaps();
  ...
}

Reply:
resetMaps is implemented all over the place. I really dont think you should be messing with this. 

As far as I can tell, the only think you really need to do is change the deactivate() and activate() methods to not unsubscribe from sensor readings (https://github.com/ros-planning/navigation2/blob/master/nav2_costmap_2d/plugins/obstacle_layer.cpp#L634) and then register the callbacks with their `is_active_[subscriber]` bool (https://github.com/ros-planning/navigation2/blob/master/nav2_costmap_2d/plugins/obstacle_layer.cpp#L207). Callback will be called but then the value is immediately passed on in the callback using that bool.

Right now you don't actually stop data from coming in and being processed on reset. This has a number of unintended consequences.

Reply:
I don't think it is that complicated. resetMaps is implemented in costmap2D. In costmap2D it used is when initializing and resizing cost maps. Other than that only the obstacle layer and voxel layer uses to clear their costmaps.

All it does is setting the costmap values to some default values.

```
void Costmap2D::resetMaps()
{
  std::unique_lock<mutex_t> lock(*access_);
  memset(costmap_, default_value_, size_x_ * size_y_ * sizeof(unsigned char));
}
```

--- 

> This has a number of unintended consequences.

Could you give an example? I am just trying to understand how to solve fix this issue without introducing new issues.


>  As far as I can tell, the only thing you really need to do is change the deactivate() and activate() methods to not unsubscribe from sensor readings (https://github.com/ros-planning/navigation2/blob/master/nav2_costmap_2d/plugins/obstacle_layer.cpp#L634) and then register the callbacks with their is_active_[subscriber] bool (https://github.com/ros-planning/navigation2/blob/master/nav2_costmap_2d/plugins/obstacle_layer.cpp#L207). Callback will be called but then the value is immediately passed on in the callback using that bool.

I will give this solution a try tomorrow.

Reply:
> resetMaps is implemented all over the place. I really dont think you should be messing with this (deleting in some places, but not others, reset doesnt seem aligned with all the plugins...).

resetMaps is part of the costmap2d class interface, not the layer interface. I expect that is why it doesn't line up exactly with the plugins. I'm not sure all the plugins inherit from both layer and costmap2d.

Every plugin needs a reset and a clearMap method since that's part of the layer interface. During those methods they may need to call resetMaps if they depend on an underlying costmap2d object.

> As far as I can tell, the only thing you really need to do is change the deactivate() and activate() methods to not unsubscribe from sensor readings

I believe activate and deactive are used by the lifecycle transitions. They need to do subscribing and unsubscribing to go through the lifecycle stages. I think it is better to just remove them from the clear costmap flow, which, I believe is what @mlherd is doing with this PR.

> Right now you don't actually stop data from coming in and being processed on reset. This has a number of unintended consequences.

If I understand your concern correctly, I believe this is safe. There are two threads in costmap, the map update thread and the main spin loop. The clearEntireCostmap service is called on the main spin loop, so no other messages can get processed until it completes. The update loop is block by a mutex taken during the processing of the service request, before the individual plugins are called to clear.